pkg/uprobetracer/tracer.go

```diff
@@ -0,0 +1,245 @@
+// Copyright 2024 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// TODO: The current implementation doesn't consider any filtering
+
+// We hold fd(s) of the executables, so we can use `/proc/self/fd/...` for attaching, it is used to avoid fd-reusing
+// We open the executables at attachContainer event, so if the file changed before the ebpf program loads,
+// we will still attach to the origin executable.
+
+// We are not maintaining ebpf.collection or perf-ring buffer by ourselves, it's hold by the parent tracer
+
+// All interfaces should hold lock, and inner functions do not
+
+package uprobetracer
+
+import (
+	"fmt"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/link"
+	"golang.org/x/sys/unix"
+
+	containercollection "github.com/inspektor-gadget/inspektor-gadget/pkg/container-collection"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/utils/host"
+)
+
+type inodeKeeper struct {
+	// if an inodeKeeper exists, `file` must be available, but `link` may be nil
+	// so if a inode is proved to not having right symbol, we can skip it for the next time
+	counter int
+	file    *os.File
+	link    link.Link
+}
+
+func (t *inodeKeeper) close() {
+	t.file.Close()
+	if t.link != nil {
+		t.link.Close()
+	}
+}
+
+type Tracer[Event any] struct {
+	progName       string
+	progType       string
+	attachFilePath string
+	attachSymbol   string
+	prog           *ebpf.Program
+
+	ctrPid2FsId    map[uint32]unix.Fsid
+	inodeRefCount  map[unix.Fsid]*inodeKeeper
```

Comment:
Could you add some comments explaining the purpose of this map and why you use `unix.Fsid` as a key?

Ditto for other maps

Reply:
Sorry, I misread the document. I thought the second value in the Fsid structure was the inodeID.